src/matrix_validator/validator_purepython.py

```diff
@@ -0,0 +1,697 @@
+"""Pure python-based validator implementation."""
+
+import csv
+import os
+import re
+import sys
+from collections import Counter, defaultdict
+
+from tqdm import tqdm
+
+from matrix_validator.validator import Validator
+
+# Increase max field size
+csv.field_size_limit(sys.maxsize)
+
+#############################
+# Configuration/Constants
+#############################
+
+# Required columns in edges file:
+REQUIRED_EDGE_COLUMNS = [
+    "subject",
+    "predicate",
+    "object",
+    "primary_knowledge_source",
+    "aggregator_knowledge_source",
+    "knowledge_level",
+    "agent_type",
+]
+
+# Required columns in nodes file:
+REQUIRED_NODE_COLUMNS = ["id", "category"]
+
+# A set of ALL valid KGX columns for edges (expand as needed)
+VALID_KGX_EDGE_COLUMNS = {
+    "publications",
+    "publications_info",
+    "kg2_ids",
+    "qualified_predicate",
+    "qualified_object_aspect",
+    "qualified_object_direction",
+    "domain_range_exclusion",
+    "id",
+    ":TYPE",
+    ":START_ID",
+    ":END_ID",
+}.union(REQUIRED_EDGE_COLUMNS)
+
+# A set of ALL valid KGX columns for nodes
+VALID_KGX_NODE_COLUMNS = {
+    "name",
+    "all_names",
+    "all_categories",
+    "iri",
+    "description",
+    "equivalent_curies",
+    "publications",
+    ":LABEL",
+}.union(REQUIRED_NODE_COLUMNS)
+
+# Possible rule names that may appear in the violations dictionary.
+# This is used to report "no violation" if they aren't triggered.
+ALL_POSSIBLE_RULES = [
+    "Naming convention violation",
+    "Line endings violation",
+    "Missing header row",
+    "Empty header column name",
+    "Duplicate header column name",
+    "Invalid KGX column (but optional ones present)",
+    "Missing required column",
+    "Inconsistent columns",
+    "Edge subject not found in node IDs",
+    "Edge object not found in node IDs",
+    "Leading or trailing whitespace",
+    "Embedded TAB character",
+    # etc., for any other rule names you have
+]
+
+# Set of biolink prefixes (obtained Jan 29 2025 for Biolink model 4.2.6-rc2 via https://w3id.org/biolink/biolink-model-prefix-map.json)
+PREFIXES = {
```

Comment:
Might make sense to pull the biolink-model.yaml file in from https://github.com/biolink/biolink-model/blob/master/biolink-model.yaml.  Or better yet, include that project as a github submodule?

Reply:
We will. This is David M's validation script verbatim. It wont stay like this - I just needed something reasonably complete to build a full cycle.